### DISCUSSION TITLE
Nix build fails on Raspberry Pi running NixOS 23.11

### DISCUSSION THREAD

**Discussion by cor** [ghostty-org/ghostty#2860] [Q&A] [answered]
Running 
```
nix build "git+ssh://git@github.com/ghostty-org/ghostty?ref=refs/heads/main&rev=e20b27de848afaa167c70d838a2f98f28c6ac0b5"
```

yields 

```
error: builder for '/nix/store/87p5fhqq4jj5f6lp8hmhck8ajybjbzkz-ghostty-0.1.0.drv' failed with exit code 1;
       lathread 326 panic: reached unreachable code
       > Panicked during a panic. Aborting.
       > error: the following build command crashed:
       > /build/source/.zig-cache/o/c06f91bee3d22324ab360943bf0bb89f/build /nix/store/2cb5q6rp7qbvq46l3rq17glkrncsa4di-zig-0.13.0/bin/zig /build/source /build/source/.zig-cache /build/tmp.lumMibFHq1 --seed 0x33548109 -Z89673cd69e052791 -Dcpu=baseline -Doptimize=ReleaseFast -Dversion-string=0.1.0-e20b27d-nix
       For full logs, run 'nix log /nix/store/87p5fhqq4jj5f6lp8hmhck8ajybjbzkz-ghostty-0.1.0.drv'.
```

`nix log /nix/store/87p5fhqq4jj5f6lp8hmhck8ajybjbzkz-ghostty-0.1.0.drv` output:

```
Running phase: unpackPhase
unpacking source archive /nix/store/4948kxsl3j9s74iq3g9d97mshw1p218z-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
zig build flags: -Dcpu=baseline -Doptimize=ReleaseFast -Dversion-string=0.1.0-e20b27d-nix
```

System info:
- Raspberry Pi 5 8GB
- `nixos-version` => 23.11.20240629.4a1e673 (Tapir)
- `github.com/cor/nixos-config/6b2d43bc0f5deb447275abc4eb9e2326531eeba4` nixosConfiguration `raspberry-pi`

Tried this in the last ~20 commits from current tip, all yield the same "Panicked during a panic. Aborting."

**Comment by cor** [comment]
Update: fails with the same error on NixOS 24.05

**Reply by cor** [reply]
Notably, the exact same build does succeed on both of my Macs, both also `aarch64-linux`, both also running NixOS 24.05 (in fact, they're built [from the same flake](github.com/cor/nixos-config))

**Comment by rhodes-b** [comment]
This is expected its an upstream issue with Zig detecting page sizes
https://github.com/ziglang/zig/pull/20511

unfortunately right now you have to patch zig yourself
https://github.com/ziglang/zig/blob/8c232922bdf21d9936db94447da21ac0dd9ec319/lib/std/mem.zig#L17
by changing this line to 16 * 1024 (16k page size vs 4k which is normal for x86-64)

**Reply by adamcstephens** [reply]
Isn't raspberry pi already at 4k? I thought it was only Apple running at 16k.

**Reply by rhodes-b** [reply]
No you can see the same issue here in the discord RaspbianOS
https://discord.com/channels/1005603569187160125/1287827296329142383

That thread does say specifically a pi 5 so maybe the old ones have 4k pages

**Reply by pluiedev** [reply, marked as answer]
This will be fixed in Zig 0.14, which means Ghostty should be able to be built from source without any hacks on Asahi soon :)

**Comment by jvoisin** [comment]
I have the same issue on Asahi Linux on an Apple M2:

```
jvoisin@chernabog 15:53 (main) ~/dev/ghostty zig build
thread 316845 panic: reached unreachable code
Panicked during a panic. Aborting.
error: the following build command crashed:
/home/jvoisin/dev/ghostty/.zig-cache/o/eb068cfe45fd6c8ce06185b729d2883c/build /usr/bin/zig /home/jvoisin/dev/ghostty /home/jvoisin/dev/ghostty/.zig-cache /home/jvoisin/.cache/zig --seed 0xc19051f1 -Z329352b1d7ccd6fa
[1]
jvoisin@chernabog 15:55 (main) ~/dev/ghostty 
```

The workaround specified [here](https://github.com/ziglang/zig/issues/16331#issuecomment-1630898319) worked.

(If only for people landing here after some Google-fu)
